shared_stack.bash

```diff
@@ -0,0 +1,214 @@
+  [ -d "$dir" ] && continue
+  $dryrun mkdir -p "${dir}.$tmp" || continue
+  (
+     set -e
+     $dryrun cd "${dir}.$tmp"
+     $dryrun bash "$STATE/lsstinstall" -T "$tag" -p "$ROOT"/conda -S
+     set +x
+     $dryrun source loadLSST.sh
+     set -x
+     # Install the developer add-on packages if needed
+     # When rubin-env-developer becomes available, replace with this:
+     # conda list --json | grep rubin-env-developer > /dev/null \
+     #   || mamba install rubin-env-developer
+     if [ -n "$dryrun" ]; then
+       echo "conda list --json | grep mypy > /dev/null || mamba install ..."
+     else
+       # shellcheck disable=SC2046
+       conda list --json | grep mypy > /dev/null \
+         || mamba install -y -c conda-forge --no-update-deps $(cat "$STATE/developer.txt")
+     fi
+     $dryrun eups distrib install -t "$tag" "$PRODUCT"
+
+     # Rewrite load scripts and write useful files
+     $dryrun cp loadLSST.sh tmp.loadLSST.sh
+     if [ -n "$dryrun" ]; then
+       echo "echo . $ROOT/loadLSST.sh ... > loadLSST.sh"
+       echo "echo conda activate ... > switch.sh"
+       echo "echo ... > env_name"
+     else
+       echo "LSST_CONDA_ENV_NAME=$LSST_CONDA_ENV_NAME . \"$ROOT/loadLSST.sh\"" > loadLSST.sh
+       echo "conda activate $LSST_CONDA_ENV_NAME" > switch.sh
+       echo "$LSST_CONDA_ENV_NAME" > env_name
+     fi
+  ) && [ ! -d "$dir" ] && $dryrun mv "${dir}.$tmp" "$dir"
+  # Clean up if we failed or lost the race
+  [ -d "${dir}.$tmp" ] && rm -rf "${dir}.$tmp"
+done
+rm $tmp/tags
+
+# List the installed tags
+(
+  cd "$ROOT/tag"
+  ls -dr d_* > "$STATE/$tmp/d_list" || true
+  ls -dr w_* > "$STATE/$tmp/w_list" || true
+)
+
+# Make the latest links if needed
+for type in d w; do
+  [ -f "$tmp/${type}_list" ] || continue
+  latest=$(head -n 1 "$tmp/${type}_list")
+  link="$ROOT/${type}_latest"
+  if [[ ! -L "$link" || "$(readlink "$link")" != "$ROOT/tag/$latest" ]]; then
+    $dryrun rm -f "$link" && $dryrun ln -s "$ROOT/tag/$latest" "$link"
+  fi
+done
+# Update loadLSST.sh if needed
+tmp_load="$ROOT/w_latest/tmp.loadLSST.sh"
+if [ -n "$dryrun" ]; then
+  echo "[ -f \"$tmp_load\" ] && mv \"$tmp_load\" \"$ROOT/loadLSST.sh\""
+elif [ -f "$tmp_load" ] && ! diff -q "$ROOT/loadLSST.sh" "$tmp_load" > /dev/null 2>&1; then
+  [ -f "$ROOT/loadLSST.sh" ] && mv "$ROOT/loadLSST.sh" "$ROOT/loadLSST.sh.bak"
+  mv "$tmp_load" "$ROOT/loadLSST.sh"
+fi
+$dryrun rm -f "$ROOT"/tag/*/tmp.loadLSST.sh
+# Update current tag to latest weekly
+# Doesn't handle all cases (an old weekly is installed as the latest in an
+# even older environment), but should be good enough
+if [ -f "$tmp/w_list" ]; then
+  (
+    latest=$(head -n 1 "$tmp/w_list")
+    set -x
+    $dryrun source "$ROOT/tag/$latest/loadLSST.sh"
+    $dryrun eups tags --clone="$latest" current
+  )
+fi
+
+#########
+# Cleanup
+#########
+
+for type in d w; do
+  [ -f "$tmp/${type}_list" ] || continue
+  keep=KEEP_$type
+  for tag_del in $(tail -n "+${!keep}" "$tmp/${type}_list"); do
```

Comment:
Isn't there some time frame where this list could be generated containing some stack, while something else is already running, such that the list is populated, but the things is removed before the body of the list is processed?

Reply:
Only one process can successfully move the tag directory to `del_dir`.  Any others will add multiple copies to `tags.deleted` but otherwise won't try to do anything else.